Fix #1

piponazo opened the pull request:
@danimtb this should address all the comments you made in #1.

I have some questions:
* What does it mean the "clean" checkbox?
* You said that "There are missing packages for Visual Studio 12, 14 and 15" but the only missing ones were the ones for Visual Studio 12. Unless I am missing something :sweat_smile: 

Please, let me know if you see something which could be improved.

piponazo committed:
Use dictionary for default options

piponazo committed:
Use cmake.definitions

piponazo committed:
Package the library license

piponazo committed:
Update travis and appveyor scripts

piponazo committed:
Add missing VS2012 images

piponazo committed:
Update expat to stable

danimtb commented on conanfile.py:
```diff
@@ -75,3 +76,6 @@ def package_info(self):
         self.cpp_info.debug.libs = ["exiv2d"]
         self.cpp_info.release.libs = ["exiv2"]
         self.cpp_info.libdirs = ['lib']  # Directories where libraries can be found
```
FYI: This is the default value of `includedirs`

danimtb replied:
FYI: This is the default value of `libdirs`

danimtb commented on conanfile.py:
```diff
@@ -23,50 +23,51 @@ class Exiv2Conan(ConanFile):
-        cmake.configure(source_dir="../exiv2", build_dir="build", defs=cmake_args)
+        cmake.configure(source_dir="../exiv2", build_dir="build")
         cmake.build()
         cmake.install()
```
FYI: Some people do the install in the package method (cmake.deifntions + cmake.configure() + cmake.install) to be able to use package development commands independently (conan build + conan package).

piponazo replied:
I read about this few months ago and It's good to know, however I never had the need to go so far. I will not implement this change for the moment.

danimtb commented on conanfile.py:
```diff
@@ -23,50 +23,51 @@ class Exiv2Conan(ConanFile):
+        cmake.definitions["EXIV2_BUILD_PO"] = "OFF"
+        cmake.definitions["EXIV2_ENABLE_XMP"] = self.options.xmp
+        cmake.definitions["EXIV2_ENABLE_PNG"] = self.options.png
+        cmake.definitions["CMAKE_INSTALL_PREFIX"] = self.package_folder
```
This is already done by the helper, see https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#definitions

danimtb commented on conanfile.py:
```diff
@@ -23,50 +23,51 @@ class Exiv2Conan(ConanFile):
+            cmake.definitions['EXIV2_ENABLE_WIN_UNICODE'] = self.options.unicode
 
-        cmake.configure(source_dir="../exiv2", build_dir="build", defs=cmake_args)
+        cmake.configure(source_dir="../exiv2", build_dir="build")
```
Use `source_folder` and `build_folder` instead. These parameters are deprecated https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#configure

danimtb reviewed (changes requested):
Thanks a lot for the update! Did some minor comments 😄

danimtb commented:
Sorry, I forgot to tick the "clean" checkbox. It usually means that the recipe does too many complex things or the code is nor very readable/maintainable (not your case).

Seems there are some missing packages of  Expat/2.2.6@pix4d/stable for clang 7, you would need to add those in its Travis matrix. There was also a random connection issue with Bintray, just realunch the job.

Please, take a look at the test_package recipe too. It will need an update on the cmake helper (source_dir/build_dir) and defining the user and channel as well as the requirement is no longer needed https://github.com/piponazo/conan-exiv2/blob/5cbfe04df7842264cafd7662fc234161775b13aa/test_package/conanfile.py#L9

Another final comment is the build type of this package. Currently it seems that the package builds by default both Release and Debug versions, so there is no need to create a package for each of them. However, I think the setting needs to be there in case someone wants the build type as "MinSizeRel" or "RelWithDebInfo". I this case you could tune up the package_id with the following:

```
def package_id(self):
    if self.settings.build_type == "Release" or self.settings.build_type == "Debug":
        self.info.settings.build_type = "DebugRelease"  # The value doesn't matter, only being the same for Debug and Release
```

You can find more information here https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#defining-a-custom-package-id

Hope it makes sense and thanks a lot for addressing the points in the review 😄

piponazo commented:
> Seems there are some missing packages of Expat/2.2.6@pix4d/stable for clang 7

Yes, I noticed. I am also updating that recipe 😉 . I will try to apply the same improvements there that you commented here.

> Another final comment is the build type of this package. Currently it seems that the package builds by default both Release and Debug versions, so there is no need to create a package for each of them.

😲 I did not know about that `DebugRelease` thing ... I will double check. 

> Hope it makes sense and thanks a lot for addressing the points in the review 😄

Thanks to you for the detailed review. I am glad to learn to write better recipes 👍

piponazo committed:
Remove lines which were using default values

piponazo committed:
Use source_folder and build_folder

piponazo commented:
Yesterday I implemented most of the changes you proposed. Just to comments more:

> Please, take a look at the test_package recipe too. It will need an update on the cmake helper (source_dir/build_dir) and defining the user and channel as well as the requirement is no longer needed

It seems that you were looking to the version in master with is an old version of the recipe. Those things are fixed in this branch. 

> Currently it seems that the package builds by default both Release and Debug versions ...

I understood later what you meant with this, and it is not the case. We only compile one version of the library depending on the `CMAKE_BUILD_TYPE` value.

Now all the CI jobs look fine, although not the ones you can see in the CI checks in this PR since I have been triggering some jobs manually.

danimtb commented:
Sorry, my fault. I looked at the master branch as you figured out.

Regarding the build type, I thought it was building both release an debug types regarding the package_info https://github.com/piponazo/conan-exiv2/blob/a6992125462053637e81411019db0a2a56bb13e8/conanfile.py#L73 but seems it is not the case 🙃
